govcd/openapi_endpoints_unit_test.go

```diff
@@ -34,40 +44,60 @@ func TestClient_getOpenApiHighestElevatedVersion(t *testing.T) {
-			endpoint:          types.OpenApiPathVersion1_0_0 + types.OpenApiEndpointNsxtNatRules,
-			wantVersion:       "34.0",
-			wantErr:           false, // NAT minimum requirement is version 34.0
+			name:                        "VCD_minimum_required_version_only",
+			supportedVersions:           renderSupportedVersions([]string{"28.0", "29.0", "30.0", "31.0", "32.0", "33.0", "34.0"}),
+			endpoint:                    types.OpenApiPathVersion1_0_0 + types.OpenApiEndpointNsxtNatRules,
+			wantVersion:                 "34.0",
+			wantErr:                     false, // NAT minimum requirement is version 34.0
+			overrideClientMinApiVersion: "34.0",
 		},
 		{
 			name: "VCD_minimum_required_version_only_unordered",
 			// Explicitly pass in unordered VCD API supported versions to ensure that ordering and matching works well
-			supportedVersions: renderSupportedVersions([]string{"33.0", "34.0", "30.0", "31.0", "32.0", "28.0", "29.0"}),
-			endpoint:          types.OpenApiPathVersion1_0_0 + types.OpenApiEndpointNsxtNatRules,
-			wantVersion:       "34.0",
-			wantErr:           false, // NAT minimum requirement is version 34.0
+			supportedVersions:           renderSupportedVersions([]string{"33.0", "34.0", "30.0", "31.0", "32.0", "28.0", "29.0"}),
+			endpoint:                    types.OpenApiPathVersion1_0_0 + types.OpenApiEndpointNsxtNatRules,
+			wantVersion:                 "34.0",
+			wantErr:                     false, // NAT minimum requirement is version 34.0
+			overrideClientMinApiVersion: "34.0",
+		},
+		{
+			name:              "VCD_version_higher_than_elevated_version_entries",
+			supportedVersions: renderSupportedVersions([]string{"37.0", "37.1", "37.2", "37.3", "38.0", "38.1", "39.0"}),
+			endpoint:          types.OpenApiPathVersion1_0_0 + types.OpenApiEndpointFirewallGroups,
+			wantVersion:       minVcdApiVersion,
+			wantErr:           false,
 		},
 	}
 
 	// Generate unit tests for each defined endpoint in endpointMinApiVersions which does not have an elevated
 	// version entry in endpointElevatedApiVersions.
-	// Always expect to get minimal version returned without error
+	// Expect to get minimal supported version returned without error
 	for endpointName, minimumRequiredVersion := range endpointMinApiVersions {
 		// Do not create a test case for those endpoints which explicitly have elevated versions defined in
 		// endpointElevatedApiVersions
 		if _, hasEntry := endpointElevatedApiVersions[endpointName]; hasEntry {
 			continue
 		}
 
+		wantVersion := minimumRequiredVersion
+		semverWantVersion, err := semver.NewSemver(wantVersion)
+		if err != nil {
+			t.Fatalf("error parsing 'singleElevatedApiVersion': %s", err)
+		}
+
+		if semverWantVersion.LessThan(semverMinVcdApiVersion) {
+			semverMinVcdApiVersionSegments := semverMinVcdApiVersion.Segments()
+			wantVersion = fmt.Sprintf("%d.%d", semverMinVcdApiVersionSegments[0], semverMinVcdApiVersionSegments[1])
```

Comment:
Suggestion for a log message when these exception conditions occur here and below.

Reply:
Good point. This is a test so I added a few additional messages for those who run tests with verbose mode. But I also added extra logging message to the core function as well.

Reply:
Done.